plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -200,7 +195,121 @@ private static void handleDynamicResourceAllocation(String[] argArray) {
           argArray[++i] = null;
         }
       }
+      // If dynamic allocation is enabled, make sure application is scheduled in right queue
+      argArray = handleQueueEnforcement(argArray);
     }
+    return argArray;
+  }
+
+  /**
+   * This method is used to enforce queue for Spark application. Rules are explained below.
+   * a) If dynamic resource allocation is enabled for selected spark version and application requires large container
+   *    then schedule it into default queue by a default conf(spark.yarn.queue) in spark-defaults.conf.
+   * b) If dynamic resource allocation is enabled for selected spark version and application requires small container
+   *    then schedule it into Org specific queue.
+   * c) If dynamic resource allocation is disabled for selected spark version then schedule application into default
+   *    queue by a default conf(spark.yarn.queue) in spark-defaults.conf.
+   * @param argArray
+   * @return
+   */
+  protected static String[] handleQueueEnforcement(String[] argArray) {
+    SparkConf sparkConf = getSparkProperties();
+    Configuration conf = new Configuration();
+    String executorMem = null;
+    String executorVcore = null;
+    String executorMemOverhead = null;
+    int queueParameterIndex = -1;
+    for (int i = 0; i < argArray.length; ++i) {
+      if (argArray[i] != null) {
+        if (argArray[i].equals(SparkJobArg.EXECUTOR_CORES.sparkParamName)) {
+          executorVcore = argArray[++i];
+        }
+        if (argArray[i].equals(SparkJobArg.EXECUTOR_MEMORY.sparkParamName)) {
+          executorMem = argArray[++i];
+        }
+        if (argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) && argArray[i + 1]
+            .startsWith(SPARK_EXECUTOR_MEMORY_OVERHEAD)) {
+          executorMemOverhead = argArray[i + 1].split("=")[1].trim();
+        }
+
+        if (argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) && argArray[i + 1]
+            .startsWith(SPARK_CONF_QUEUE) || argArray[i].equals(SparkJobArg.QUEUE.sparkParamName)) {
+          queueParameterIndex = i;
+        }
+      }
+    }
+
+    boolean requiredSparkDefaultQueue = false;
+    if (sparkConf.getBoolean(SPARK_CONF_DYNAMIC_ALLOC_ENABLED, false)) {
+      if (isLargeContainerRequired(executorVcore, executorMem, executorMemOverhead, conf, sparkConf)) {
+        // Case A
+        requiredSparkDefaultQueue = true;
+        logger.info(
+            "Spark application requires Large containers. Scheduling this application into default queue by a "
+                + "default conf(spark.yarn.queue) in spark-defaults.conf.");
+      } else {
+        // Case B
+        logger.info(
+            "Dynamic allocation is enabled for selected spark version and application requires small container. "
+                + "Hence, scheduling this application into Org specific queue");
+        if(queueParameterIndex == -1) {
```

Comment:
If user explicitly specify spark_default queue in the queue parameter, we should also change the queue to default.